source/slang/slang-syntax.cpp

```diff
@@ -85,6 +85,15 @@ bool SyntaxClassBase::isSubClassOfImpl(SyntaxClassBase const& super) const
     return false;
 }
 
+NodeBase* _dynamicCastImpl(NodeBase* node, SyntaxClassBase const& toClass)
```

Comment:
It's a small point, but because in effect this is doing static_cast, you might want to do

```
template <typename T>
SLANG_FORCE_INLINE T* dynamicCast(NodeBase* node)
{
    return node && node->getClass().isSubClassOfImpl(getClass<T>()) ? static_cast<T*>(node) : nullptr; 
}
SLANG_FORCE_INLINE const T* dynamicCast(const NodeBase* node)
{
    return node && node->getClass().isSubClassOfImpl(getClass<T>()) ? static_cast<const T*>(node) : nullptr; 
}
```
That using static_cast<> makes clearer, and means must be in principal convertible, and there is no need for const_cast.

Reply:
I think I agree that this could be written more compactly, and maybe without the extra function call in the `const` case, but my main goal here was to move as much of the actual logic out of the templated functions and into the non-templated `_dynamicCastImpl` routine to avoid as much template bloat as possible. Your proposal kind of goes in the other direction of assuming that force-inlining things will lead to better perf with minimal overhead at call sites.

One thing to note here is that part of why I moved things out to a worker routine in my subsequent fixup changes was that the place where these functions were being declared in `slang-syntax.h` meant that `NodeBase` wasn't yet fully defined and clang/gcc gave errors if I tried to define the body of the functions inline (e.g., `node->getClass()` isn't valid when `NodeBase` hasn't been defined yet).